velox/common/caching/SsdCache.cpp

```diff
@@ -0,0 +1,505 @@
+          verifyWrite(*entry, SsdRun(offset, size));
+        }
+        offset += size;
+        ++stats_.entriesWritten;
+        stats_.bytesWritten += size;
+      }
+    }
+    storeIndex += numWritten;
+  }
+}
+
+char* readBytes(int fd, int64_t offset, int32_t size) {
+  char* data = (char*)malloc(size);
+  pread(fd, data, size, offset);
+  return data;
+}
+
+namespace {
+int32_t indexOfFirstMismatch(char* x, char* y, int n) {
+  for (auto i = 0; i < n; ++i) {
+    if (x[i] != y[i]) {
+      return i;
+    }
+  }
+  return -1;
+}
+} // namespace
+
+void SsdFile::verifyWrite(AsyncDataCacheEntry& entry, SsdRun ssdRun) {
+  auto testData = std::make_unique<char[]>(entry.size());
+  auto rc = pread(fd_, testData.get(), entry.size(), ssdRun.offset());
+  VELOX_CHECK_EQ(rc, entry.size());
+  if (entry.tinyData()) {
+    if (0 != memcmp(testData.get(), entry.tinyData(), entry.size())) {
+      VELOX_FAIL("bad read back");
+    }
+  } else {
+    auto& data = entry.data();
+    int64_t bytesLeft = entry.size();
+    int64_t offset = 0;
+    for (auto i = 0; i < data.numRuns(); ++i) {
+      auto run = data.runAt(i);
+      auto compareSize = std::min<int64_t>(bytesLeft, run.numBytes());
+      auto badIndex = indexOfFirstMismatch(
+          run.data<char>(), testData.get() + offset, compareSize);
+      if (badIndex != -1) {
+        VELOX_FAIL("Bad read back");
+      }
+      bytesLeft -= run.numBytes();
+      offset += run.numBytes();
+      if (bytesLeft <= 0) {
+        break;
+      };
+    }
+  }
+}
+
+void SsdFile::updateStats(SsdCacheStats& stats) {
+  stats.entriesWritten += stats_.entriesWritten;
+  stats.bytesWritten += stats_.bytesWritten;
+  stats.entriesRead += stats_.entriesRead;
+  stats.bytesRead += stats_.bytesRead;
+  stats.entriesCached += entries_.size();
+  for (auto& regionSize : regionSize_) {
+    stats.bytesCached += regionSize;
+  }
+}
+
+void SsdFile::clear() {
+  std::lock_guard<std::mutex> l(mutex_);
+  entries_.clear();
+  std::fill(regionSize_.begin(), regionSize_.end(), 0);
+  writableRegions_.resize(numRegions_);
+  std::iota(writableRegions_.begin(), writableRegions_.end(), 0);
+  std::fill(regionScore_.begin(), regionScore_.end(), 0);
+}
+
+SsdCache::SsdCache(
+    std::string_view filePrefix,
+    uint64_t maxBytes,
+    int32_t numShards)
+    : filePrefix_(filePrefix),
+      numShards_(numShards),
+      groupStats_(std::make_unique<FileGroupStats>()) {
+  files_.reserve(numShards_);
+  uint64_t kSizeQuantum = numShards_ * SsdFile::kRegionSize;
+  int32_t fileMaxRegions = bits::roundUp(maxBytes, kSizeQuantum) / kSizeQuantum;
+  for (auto i = 0; i < numShards_; ++i) {
+    files_.push_back(std::make_unique<SsdFile>(
+        fmt::format("{}{}", filePrefix_, i), *this, i, fileMaxRegions));
+  }
+}
+
+SsdFile& SsdCache::file(uint64_t fileId) {
+  auto index = fileId % numShards_;
+  return *files_[index];
+}
+
+namespace {
+folly::IOThreadPoolExecutor* ssdStoreExecutor() {
```

Comment:
I think we'd want this executor to be provided by the application, not hard-coded inside the library.